## Student Performance Indicator
### Life cycle of Machine Leraning Project

- Understanding the Problem Statement 
-Data Collection
-Data Checks to Perform
-Exploratory Data Analysis 
-Data Pre-Processing
-Model Training
-Choose Best Model

### 1.  Problem Statement:
- This project understands how the student's performance(test scores) is affected by other variables such as Gender. Ethnicity Parental level of education, Lunch and Test Prepration Course.

### 2. Data Collection:
- Dataset Source-  https://www.kaggle.com/datasets/spscientist/students-performance-in-exams?datasetId=74977
- The data consists of 8 column and 1000 rows 

### 2.1 Import Data and Required Packages
- Importing Pandas, Numpy, Matplotlib,Seaborn and Warnings Library

### 2.2 Import the CSV Data as Pandas DataFrame

In [1]:
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings("ignore")

In [2]:
df = pd.read_csv('data/student.csv')

In [3]:
# Show Top 10 Records:
df.head(10)

gender race/ethnicity parental level of education         lunch  \
0  female        group D                some college      standard   
1    male        group D          associate's degree      standard   
2  female        group D                some college  free/reduced   
3    male        group B                some college  free/reduced   
4  female        group D          associate's degree      standard   
5    male        group C            some high school      standard   
6  female        group E          associate's degree      standard   
7  female        group B            some high school      standard   
8    male        group C            some high school      standard   
9  female        group C           bachelor's degree      standard   

  test preparation course  math score  reading score  writing score  
0               completed          59             70             78  
1                    none          96             93             87  
2                    none          57             76             77  
3                    none          70             70             63  
4                    none          83             85             86  
5                    none          68             57             54  
6                    none          82             83             80  
7                    none          46             61             58  
8                    none          80             75             73  
9               completed          57             69             77

In [5]:
# Shape of the Dataset
df.shape

(1000, 8)

### 2.2 Dataset information
- gender : sex of students -> (Male/female)
-race/ethnicity : ethnicity of students -> (Group A, B,C, D,E)
-parental level of education : parents' final education ->(bachelor's degree,some college,master's degree,associate's degree,high school)
-lunch : having lunch before test (standard or free/reduced)
-test preparation course : complete or not complete before test
-math score
-reading score
-writing score

### 3. Data Checks to perform
- Check Missing values
-Check Duplicates
-Check data type
-Check the number of unique values of each column
-Check statistics of data set
-Check various categories present in the different categorical column

In [6]:
# 1.Check Missing Values 
df.isna().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

In [7]:
# 2.Check Duplicates
df.duplicated().sum()

0

In [8]:
# 3.Check data types and Null

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


In [9]:
# 4.Checking the number of unique values of each column

df.nunique()

gender                          2
race/ethnicity                  5
parental level of education     6
lunch                           2
test preparation course         2
math score                     79
reading score                  71
writing score                  74
dtype: int64

In [10]:
# 5.Check statistics of dataset
df.describe()

math score  reading score  writing score
count  1000.000000    1000.000000    1000.000000
mean     67.810000      70.382000      69.140000
std      15.250196      14.107413      15.025917
min      15.000000      25.000000      15.000000
25%      58.000000      61.000000      59.000000
50%      68.000000      70.500000      70.000000
75%      79.250000      80.000000      80.000000
max     100.000000     100.000000     100.000000

## Insight
- From above description of numerical data, all means are very close to each other - between 67.8 and 70.38;
- All standard deviations are also close - between 14.1 and 15.25;
- While there is a minimum score 15 for math, for writing minimum is equal to 15 and for reading is much higher = 25

In [11]:
# 6.EXploring Data 
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group D                some college      standard   
1    male        group D          associate's degree      standard   
2  female        group D                some college  free/reduced   
3    male        group B                some college  free/reduced   
4  female        group D          associate's degree      standard   

  test preparation course  math score  reading score  writing score  
0               completed          59             70             78  
1                    none          96             93             87  
2                    none          57             76             77  
3                    none          70             70             63  
4                    none          83             85             86

In [16]:
print("Categories in 'gender' variable: ",end=" " )
print(df['gender'].unique())

print("Categories in 'race_ethnicity' variable: ",end=" ")
print(df['race/ethnicity'].unique())

print("Categories in'parental level of education' variable: ",end=" " )
print(df['parental level of education'].unique())

print("Categories in 'lunch' variable: ",end=" " )
print(df['lunch'].unique())

print("Categories in 'test preparation course' variable: ",end=" " )
print(df['test preparation course'].unique())

Categories in 'gender' variable:  ['female' 'male']
Categories in 'race_ethnicity' variable:  ['group D' 'group B' 'group C' 'group E' 'group A']
Categories in'parental level of education' variable:  ['some college' "associate's degree" 'some high school'
 "bachelor's degree" "master's degree" 'high school']
Categories in 'lunch' variable:  ['standard' 'free/reduced']
Categories in 'test preparation course' variable:  ['completed' 'none']


In [17]:
# 7.define numerical & categorical columns
numeric_features = [feature for feature in df.columns if df[feature].dtype != 'O']
categorical_features = [feature for feature in df.columns if df[feature].dtype == 'O']

# print columns
print('We have {} numerical features : {}'.format(len(numeric_features), numeric_features))
print('\nWe have {} categorical features : {}'.format(len(categorical_features), categorical_features))

We have 3 numerical features : ['math score', 'reading score', 'writing score']

We have 5 categorical features : ['gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course']


In [19]:
# 8.Adding columns for 'Toatal Score' and 'Average'

df['total score'] = df['math score'] + df['reading score'] + df['writing score']
df['average'] = df['total score']/3
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group D                some college      standard   
1    male        group D          associate's degree      standard   
2  female        group D                some college  free/reduced   
3    male        group B                some college  free/reduced   
4  female        group D          associate's degree      standard   

  test preparation course  math score  reading score  writing score  \
0               completed          59             70             78   
1                    none          96             93             87   
2                    none          57             76             77   
3                    none          70             70             63   
4                    none          83             85             86   

   total score    average  
0          207  69.000000  
1          276  92.000000  
2          210  70.000000  
3          203  67.666667  
4          254  84.666667

In [20]:
reading_full = df[df['reading score'] == 100]['average'].count()
writing_full = df[df['writing score'] == 100]['average'].count()
math_full = df[df['math score'] == 100]['average'].count()

print(f'Number of students with full marks in Maths: {math_full}')
print(f'Number of students with full marks in Writing: {writing_full}')
print(f'Number of students with full marks in Reading: {reading_full}')

Number of students with full marks in Maths: 9
Number of students with full marks in Writing: 13
Number of students with full marks in Reading: 10


In [22]:
reading_less_30 = df[df['reading score'] <= 30]['average'].count()
writing_less_30 = df[df['writing score'] <= 30]['average'].count()
math_less_30 = df[df['math score'] <= 30]['average'].count()

print(f'Number of students with less than 30 marks in Maths: {math_less_30}')
print(f'Number of students with less than 30 marks in Writing: {writing_less_30}')
print(f'Number of students with less than 30 marks in Reading: {reading_less_30}')

Number of students with less than 30 marks in Maths: 11
Number of students with less than 30 marks in Writing: 7
Number of students with less than 30 marks in Reading: 4


### Insights
- From above values we get students have performed the worst in Maths
- Best performance is in reading section

### 4. Exploring Data (Visualization)
#### 4.1 Visualize average score distribution to make some conclusion.
- Histogram
- Kernel Distribution Function (KDE)

In [ ]:
# 4.1.1 Histogram & KDE 

fig, axs = plt.subplots(1, 2, figsize=(15, 7))
plt.subplot(121)
sns.histplot(data=df, x='average', bins=30, kde=True, color='g')
plt.subplot(122)
sns.histplot(data=df, x='average', kde=True, hue='gender')
plt.show()

In [ ]:
fig, axs = plt.subplots(1, 2, (figsize=(15, 7))
plt.subplot(121) 
sns.histplot(data=df, x='total score', bins=30, kde=True, color='g')
plt.subplot(122)
sns.histplot(data=df, x='tatal score', kde=True, hue='gender')
plt.show()

### Insights
- Female students tend to perform well then male students.

In [ ]:
plt.subplots(1,3. figsize=(25,6))
plt.subplot(141)
sns.histplot(data=df, x='average',kde=True,hue='lunch')
plt.subplot(142)
sns.histplot(data=df[df.gender=='female'],x='average',kde=True,hue='lunch')
plt.subplot(143)
sns.histplot(data=df[df.gender=='male'],x='average',kde=True,hue='lunch')
plt.show()

### Insights
- Standard lunch helps perform well in exams.
- Standard lunch helps perform well in exams be it a male or a female.

In [ ]:
plt.subplots(1,3,figsize=(25,6))
plt.subplot(141)
ax =sns.histplot(data=df,x='average',kde=True,hue='parental level of education')
plt.subplot(142)
ax =sns.histplot(data=df[df.gender=='male'],x='average',kde=True,hue='parental level of education')
plt.subplot(143)
ax =sns.histplot(data=df[df.gender=='female'],x='average',kde=True,hue='parental level of education')
plt.show()

### Insights
- In general parent's education don't help student perform well in exam.
- 2nd plot shows that parent's whose education is of associate's degree or master's degree their male child tend to perform well in exam
- 3rd plot we can see there is no effect of parent's education on female students.

In [ ]:
plt.subplots(1,3,figsize=(25,6))
plt.subplot(141)
ax =sns.histplot(data=df,x='average',kde=True,hue='race/ethnicity')
plt.subplot(142)
ax =sns.histplot(data=df[df.gender=='female'],x='average',kde=True,hue='race/ethnicity')
plt.subplot(143)
ax =sns.histplot(data=df[df.gender=='male'],x='average',kde=True,hue='race/ethnicity')
plt.show()

### Insights
- Students of group A and group B tends to perform poorly in exam.
- Students of group A and group B tends to perform poorly in exam irrespective of whether they are male or female

### 4.2 Maximum score  of students in all three subjects.

In [ ]:
plt.figure(figsize=(18,8))
plt.subplot(1, 4, 1)
plt.title('MATH SCORES')
sns.violinplot(y='math score',data=df,color='red',linewidth=3)
plt.subplot(1, 4, 2)
plt.title('READING SCORES')
sns.violinplot(y='reading score',data=df,color='green',linewidth=3)
plt.subplot(1, 4, 3)
plt.title('WRITING SCORES')
sns.violinplot(y='writing score',data=df,color='blue',linewidth=3)
plt.show()

### Insights
- From the above three plots its clearly visible that most of the students score in between 60-80 in Maths whereas in reading and writing most of them score from 50-80

### 4.3 Multivariate analysis using pieplot